File: backend/app/api/v1/addresses.py
from typing import Annotated

from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.api.exceptions import (
    BadRequestException,
    ConflictException,
    ForbiddenException,
    NotFoundException,
)
from app.models.user import User
from app.schemas.address import AddressCreate, AddressResponse, AddressUpdate
from app.services.address_service import AddressService

address_router = APIRouter()

DbSession = Annotated[Session, Depends(get_db)]
UserDb = Annotated[User, Depends(get_current_user)]


def get_address_service(db: DbSession) -> AddressService:
    return AddressService(db)


def _traduzir_value_error(exc: ValueError) -> BadRequestException:
    """Traduz os ``ValueError`` do service em erros HTTP (senão viram 500)."""
    msg = str(exc)

    if "not owned by user" in msg:
        return ForbiddenException(msg)

    if "already exists" in msg:
        return ConflictException(msg, code="ADDRESS_ALREADY_EXISTS")

    if "No address" in msg:
        return NotFoundException(msg, code="ADDRESS_NOT_FOUND")

    return BadRequestException(msg)


@address_router.post(
    "/create",
    response_model=AddressResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Cria um novo endereço para o usuário autenticado",
)
def create_address(
    data: AddressCreate, user: UserDb, db: DbSession
) -> AddressResponse:
    try:
        return get_address_service(db).create(data, user.id)
    except ValueError as exc:
        raise _traduzir_value_error(exc) from exc


@address_router.get(
    "/list",
    response_model=list[AddressResponse],
    summary="Lista os endereços do usuário autenticado",
)
def list_addresses(user: UserDb, db: DbSession) -> list:
    try:
        return get_address_service(db).get_by_user_id(user.id)
    except ValueError as exc:
        raise _traduzir_value_error(exc) from exc


@address_router.get(
    "/get/{address_id}",
    response_model=AddressResponse,
    summary="Busca um endereço pelo ID",
)
def get_address(
    address_id: int, user: UserDb, db: DbSession
) -> AddressResponse:
    try:
        return get_address_service(db).get_by_id(user.id, address_id)
    except ValueError as exc:
        raise _traduzir_value_error(exc) from exc


@address_router.get(
    "/default",
    response_model=AddressResponse,
    summary="Busca o endereço padrão do usuário",
)
def get_default_address(user: UserDb, db: DbSession) -> AddressResponse:
    try:
        return get_address_service(db).get_actual_address_default(user.id)
    except ValueError as exc:
        raise _traduzir_value_error(exc) from exc


@address_router.get(
    "/zip/{zip_code}",
    response_model=AddressResponse,
    summary="Busca um endereço pelo CEP",
)
def get_address_by_zip_code(
    zip_code: str, user: UserDb, db: DbSession
) -> AddressResponse:
    try:
        return get_address_service(db).get_by_zip_code(user.id, zip_code)
    except ValueError as exc:
        raise _traduzir_value_error(exc) from exc


@address_router.patch(
    "/default/set/{address_id}",
    response_model=AddressResponse,
    summary="Define um endereço como padrão",
)
def set_default_address(
    address_id: int, user: UserDb, db: DbSession
) -> AddressResponse:
    try:
        return get_address_service(db).set_default(user.id, address_id)
    except ValueError as exc:
        raise _traduzir_value_error(exc) from exc


@address_router.patch(
    "/update/{address_id}",
    response_model=AddressResponse,
    summary="Atualiza um endereço",
)
def update_address(
    address_id: int, data: AddressUpdate, user: UserDb, db: DbSession
) -> AddressResponse:
    try:
        return get_address_service(db).update(address_id, data, user.id)
    except ValueError as exc:
        raise _traduzir_value_error(exc) from exc


@address_router.delete(
    "/delete/{address_id}",
    response_model=AddressResponse,
    summary="Exclui um endereço",
)
def delete_address(
    address_id: int, user: UserDb, db: DbSession
) -> AddressResponse:
    try:
        return get_address_service(db).delete(address_id, user.id)
    except ValueError as exc:
        raise _traduzir_value_error(exc) from exc

File: frontend/Papiro/src/api/addresses.js
import api from './client'

/**
 * Endpoints de endereços.
 *
 * Contrato atual do backend (FastAPI):
 *   POST   /addresses/create            -> cria (201). Corpo: AddressCreate.
 *   GET    /addresses/list              -> lista os endereços do usuário.
 *   GET    /addresses/default           -> endereço padrão.
 *   GET    /addresses/zip/{zip_code}    -> busca por CEP.
 *   PATCH  /addresses/default/set/{id}  -> torna padrão.
 *   PATCH  /addresses/update/{id}       -> atualiza.
 *   DELETE /addresses/delete/{id}       -> exclui.
 *
 * AUTENTICAÇÃO: todas as rotas exigem Bearer token e derivam o `user_id` do
 * usuário autenticado (`Depends(get_current_user)`). O `user_id` NÃO vai mais
 * na query nem no corpo — o interceptor do client anexa o token sozinho.
 */
export const addressService = {
  /**
   * Cria um endereço de entrega para o usuário autenticado.
   * @param {{ street: string, number: string, complement?: string|null, neighborhood: string, city: string, state: string, zip_code: string, is_default?: boolean }} endereco
   */
  async criar(endereco) {
    const { data } = await api.post('/addresses/create', endereco)
    return data
  },

  /** Lista os endereços do usuário autenticado. */
  async listar() {
    try {
      const { data } = await api.get('/addresses/list')
      return Array.isArray(data) ? data : []
    } catch (error) {
      if (error?.status === 404 || error?.code === 'ADDRESS_NOT_FOUND') {
        return []
      }
      throw error
    }
  },

  /** Busca o endereço padrão. */
  async padrao() {
    try {
      const { data } = await api.get('/addresses/default')
      return data
    } catch (error) {
      if (error?.status === 404 || error?.code === 'ADDRESS_NOT_FOUND') {
        return null
      }
      throw error
    }
  },

  /** Define um endereço como padrão. */
  async definirPadrao(addressId) {
    const { data } = await api.patch(`/addresses/default/set/${addressId}`)
    return data
  },

  /** Atualiza um endereço existente. */
  async atualizar(addressId, endereco) {
    const { data } = await api.patch(`/addresses/update/${addressId}`, endereco)
    return data
  },

  /** Exclui um endereço. */
  async excluir(addressId) {
    const { data } = await api.delete(`/addresses/delete/${addressId}`)
    return data
  },
}

export default addressService
